### PR TITLE
Fix #1911: Enable -XfatalWarnings where we can

### LINKED ISSUE #1911: Enable -XfatalWarnings where we can
Most projects do compile without warnings. We should enable `-XfatalWarnings`

### PR DESCRIPTION
**Pull request by gzm0**
I did not change the examples in order to not collide with another upcoming change.

### CODE REVIEW

**Comment by gzm0** [comment]
Review by @sjrd

**Comment by nicolasstucki** [comment]
LGTM

File: cli/src/main/scala/org/scalajs/cli/Scalajsld.scala [hunk 1]
```diff
@@ -163,13 +163,12 @@ object Scalajsld {
...
         newScalaJSOptimizer(semantics, options.outputMode), cp,
-        Config(
-            output = output,
-            wantSourceMap = options.sourceMap,
-            relativizeSourceMapBase = options.relativizeSourceMap,
-            bypassLinkingErrors = options.bypassLinkingErrors,
-            checkIR = options.checkIR,
-            prettyPrint = options.prettyPrint),
+        Config(output)
+            .withWantSourceMap(options.sourceMap)
```
Comment (sjrd) [inline cli/src/main/scala/org/scalajs/cli/Scalajsld.scala]: This should be 2 spaces, not 4. Same for the rest of the commit of course.

File: project/Build.scala [hunk 2]
```diff
@@ -321,7 +325,7 @@ object Build extends sbt.Build {
   )
 
   val commonToolsSettings = (
-      commonSettings ++ publishSettings
+      commonSettings ++ publishSettings ++ fatalWarningsSettings
```
Comment (sjrd) [inline project/Build.scala]: I believe the tools have warnings (pattern matching not being able to prove reachability or some other nonsense)

**Comment by gzm0** [comment]
Updated

**Comment by sjrd** [comment]
LGTM

File: project/Build.scala [hunk 3]
```diff
@@ -1036,7 +1062,7 @@ object Build extends sbt.Build {
   lazy val partest: Project = Project(
       id = "partest",
       base = file("partest"),
-      settings = commonSettings ++ Seq(
+      settings = commonSettings ++ fatalWarningsSettings ++ Seq(
```
Comment (sjrd) [inline project/Build.scala]: This is the cause of the build failure. partest sources in the tools, so it builds with warnings, unfortunately.

Comment (gzm0) [inline project/Build.scala]: I am considering to turn of analysis for the ir and the tools. IIUC it does not work anyways: We have default cases everywhere (so exhaustiveness is useless) and we fail reachability everywhere. WDYT?

Comment (sjrd) [inline project/Build.scala]: If you manage to turn it off, OK. But I think I tried to, and sometimes it doesn't really turn it off even if you ask for it.

Comment (gzm0) [inline project/Build.scala]: I can't even reproduce the warning locally... The tools compile just fine

**Comment by gzm0** [comment]
Updated so upstream stuff can continue. I'll investigate the patmat issues separately.

**Comment by sjrd** [comment]
Is this supposed to be good to merge?

**Comment by gzm0** [comment]
Yes, if it LGTY, you can merge.

**Comment by sjrd** [comment]
LGTM

**Pull request merged**